feature: new tags command

tldahlgren opened the pull request:
Fixes #22938 

This PR adds a `spack tags` command to output package tags or (available) packages with those tags.  It also ensures each package is listed in the tag cache ONLY ONCE per tag.

```
$ spack tags --help
usage: spack tags [-hia] [tag [tag ...]]

Show package tags and associated packages

positional arguments:
  tag              show packages with the specified tag

optional arguments:
  -h, --help       show this help message and exit
  -i, --installed  show information for installed packages only
  -a, --all        show packages for all available tags

Tags from known packages will be used if no tags are provided on the command
line. If tags are provided, packages with at least one will be reported.

You are not allowed to provide tags and use '--all' at the same time.


$ spack tags
30 available package tags:
    amr            finite-elements  hpc            radiuss
    build-tools    flow             hydrodynamics  simulations
    cfd            fluid            i/o            solver
    e4s            gibbs sampler    key4hep        spectral-elements
    ecp            gibbs-sampler    mcmc           storage
    ecp-apps       hep              navier-stokes  trace-tools
    ecp-proxy-app  heps             neutrino
    fem            high-order       proxy-app


$ spack tags -a
amr:
    mfem

build-tools:
    autoconf  cmake  gawk   libtool  pkg-config
    automake  flex   gmake  m4       pkgconf

[..snip..]

trace-tools:
    uftrace

$ spack tags -is
amr:
    No installed packages

build-tools:
    autoconf  automake  cmake  flex  libtool  m4  pkg-config  pkgconf

[..snip..]

trace-tools:
    No installed packages


$ spack tags ecp ecp-apps ecp-proxy-app
ecp:
    amr-wind          er       lammps     nekrs      qmcpack  shuffile
    axl               exawind  latte      nektools   rankstr  spath
    chombo            filo     nalu-wind  nwchem     redset   veloc
    ecp-data-vis-sdk  kvtree   nek5000    parsplice  rmgdft

ecp-apps:
    amr-wind  exawind  latte      nek5000  nektools  parsplice  rmgdft
    chombo    lammps   nalu-wind  nekrs    nwchem    qmcpack

ecp-proxy-app:
    amg                ember      macsio   miniqmc   nekbone     swfft
    candle-benchmarks  examinimd  miniamr  minitri   picsarlite  thornado-mini
    ecp-proxy-apps     laghos     minife   minivite  sw4lite     xsbench


$ spack tags -i radiuss ecp
ecp:
    No installed packages

radiuss:
    blt  hypre  raja  umpire  zfp
```

TODO
- [x] Add an option to limit reported tags to those of installed packages
- [x] Rename `spack.tag_index` to `spack.tag`
- [x] Move `spack.cmd.tag` functions into `spack.tag`
- [x] Remove `-t` option from `spack list`
- [x] Replace spaces embedded in tags with a hyphen

~Deferring addressing package tags with embedded spaces to a separate PR.  Do not appear to be able to use standard mechanisms for including them on the `spack tags` command line but `spack tags` will show them.~

tldahlgren committed:
feature: new tags command

alalazo reviewed (commented):
Can we rename the Python package to `spack.tag` instead of `spoack.tag_index` and move in there the logic that is currently in `spack.cmd.tag`? 

In that way we'll make the command a thin wrapper around internal API, and can code more unit-tests avoiding to use `SpackCommand` directly.

I think we also need to remove an option to the `spack list` command that becomes redundant with this PR.

tldahlgren commented:
> I think we also need to remove an option to the `spack list` command that becomes redundant with this PR.

Should the option also be removed from `spack find`?

tldahlgren committed:
Renamed tag_index.py to tag.py

tldahlgren committed:
Oops. Updated bash completion

tldahlgren commented:
Force re-running checks due to an odd `gpg` error in one.

alalazo commented:
@spackbot run pipeline

tldahlgren commented:
Note pipeline check fails for E4S stack for a known broken spec:

``Error: Pipeline generation failed due to the presence of the following specs that are known to be broken in develop``

tldahlgren commented:
@spackbot run pipeline

tldahlgren commented:
> Can we rename the Python package to `spack.tag` instead of `spoack.tag_index` and move in there the logic that is currently in `spack.cmd.tag`?
> 
> In that way we'll make the command a thin wrapper around internal API, and can code more unit-tests avoiding to use `SpackCommand` directly.
> 
> I think we also need to remove an option to the `spack list` command that becomes redundant with this PR.

@alalazo I think I made all of your requested changes.

tldahlgren commented:
@alalazo This PR also fixes a problem where the tags cache ends up with multiple entries for the same package under a tag.

tldahlgren committed:
Remove only remaining tag with embedded space

tldahlgren commented:
@alalazo Are there any more changes you want to this PR?

alalazo commented on lib/spack/spack/cmd/tags.py:
```diff
@@ -0,0 +1,90 @@
+    subparser.add_argument(
+        '-s', '--show-packages', action='store_true', default=False,
+        help="show packages for available tags when no tags are provided"
+    )
```
What do you think  about calling this:
```suggestion
    subparser.add_argument(
        '--all', action='store_true', default=False,
        help="show packages for all the available tags"
    )
```
and making it such that either this option or a list of tags can be provided from the command line?

tldahlgren replied:
IIRC I was having sub parser problems with mutual exclusion but will give it another try.

tldahlgren replied:
I don't appear (Python 3.7?) to be able to have the packages option mutually exclusive with the positional tag arguments *unless* I'm willing to support only a single tag and that was not the intention.

alalazo replied:
https://github.com/spack/spack/blob/361d98c4ec93f9b682b949a111958df38e70f46b/lib/spack/spack/cmd/tags.py#L61-L64 

This is absolutely fine, thanks!

alalazo commented on lib/spack/spack/cmd/tags.py:
```diff
@@ -0,0 +1,90 @@
+    subparser.add_argument(
+        'tag',
+        nargs='*',
+        help="find packages that provide this tag"
```
Nitpick, for consistency with other help strings:
```suggestion
        help="show packages that provide this tag"
```

alalazo commented on lib/spack/spack/cmd/tags.py:
```diff
@@ -0,0 +1,90 @@
+import spack.repo
+import spack.tag
+
+description = "list package tags and or their packages"
```
Just a suggestion (not a request):
```suggestion
description = "show tags and list their packages"
```

alalazo commented on lib/spack/spack/cmd/tags.py:
```diff
@@ -0,0 +1,90 @@
+        return
+
+    # Always show packages when tags are provided
+    show_packages = args.tag or args.show_packages
```
Should we intercept if both flags are given and print an error message in that case? Asking since:
```console
$ spack tags -s
[ shows all tags ]
$ spack tags -s ecp
ecp:
    amr-wind  arborx  cabana  chombo            er       filo    kvtree  latte      nek5000  nektools  parsplice  rankstr  rmgdft    spath
    amrex     axl     chai    ecp-data-vis-sdk  exawind  heffte  lammps  nalu-wind  nekrs    nwchem    qmcpack    redset   shuffile
```
seems inconsistent.

alalazo commented on lib/spack/spack/cmd/tags.py:
```diff
@@ -0,0 +1,90 @@
+        else:
+            tag_pkgs = spack.tag.get_tag_packages(available_tags, True, True)
+            tags = tag_pkgs.keys() if tag_pkgs else []
+            report_tags("Installed", tags)
```
Minor (I'd say not a request, but would be welcome). Can we improve the output in this particular case :
```console
$ spack tags -i
Installed package tags:
    None
```
? I think:
```console
$ spack tags -i
==> Installed packages have no tag
```
in case we have packages in the DB but without tags and:
```console
$ spack tags -i
==> No installed packages
```
in case we have an empty DB would look slightly better.

tldahlgren replied:
I was following the `spack info` precedent but can switch.

alalazo replied:
I think everything is great in the output with this single exception. Start from an empty DB then:
```console
$ spack install zlib
[ ... ]
$ spack tags -i
0 installed package tags:
    None
```
Can we easily output something like:
```console
$ spack tags -i
==> Installed packages have no tag
```
?

alalazo commented on lib/spack/spack/tag.py:
```diff
@@ -0,0 +1,136 @@
+import spack.util.spack_json as sjson
+
+
+def get_installed_packages():
```
I'd be fine with this if it is an underscore function:
```suggestion
def _get_installed_packages():
```
If it is part of the public API, it should probably be reworked and moved into `spack.environment` or `spack.store`?. Rationale for this is that the function does not deal with tags at all, just with environments and installed specs.

alalazo commented on lib/spack/spack/tag.py:
```diff
@@ -0,0 +1,136 @@
+    return [spec.name for spec in spack.store.db.query(hashes=hashes)]
+
+
+def get_tag_packages(tags, installed, skip_empty):
```
What do you think of:
```suggestion
def packages_with_tags(tags, installed, skip_empty):
```
as a name for the function?

alalazo commented on lib/spack/spack/test/cmd/tags.py:
```diff
@@ -0,0 +1,59 @@
+
+import spack.repo
+from spack.main import SpackCommand
+from spack.spec import Spec
```
Minor. Can we keep fully qualified imports:
```python
import spack.main
import spack.spec
```
?

alalazo commented on lib/spack/spack/test/cmd/tags.py:
```diff
@@ -0,0 +1,59 @@
+def test_tags_packages(install_mockery, mock_fetch, args, packages):
+    out = tags(*args)
+    for pkg in packages:
+        assert pkg in out
```
It's unclear to me what behavior we are testing. Is it just to hit lines of code in `cmd` and increase coverage?

tldahlgren replied:
Basically.  I realize two cases aren't needed so can remove.

alalazo commented on lib/spack/spack/test/tag.py:
```diff
@@ -0,0 +1,160 @@
+
+def test_tag_get_all_available(mock_packages):
+    for skip in [False, True]:
+        all_pkgs = spack.tag.get_tag_packages([], False, skip)
```
Minor. I know dynamic typing is not the best for these kind of things, but I would be consistent with the docstring:
```suggestion
        all_pkgs = spack.tag.get_tag_packages(None, False, skip)
```

alalazo commented on lib/spack/spack/tag.py:
```diff
@@ -0,0 +1,136 @@
+    """
+    tag_pkgs = collections.defaultdict(lambda: list)
+    spec_names = get_installed_packages() if installed else []
+    keys = tags or spack.repo.path.tag_index
```
Should we be more defensive here:
```suggestion
    keys = spack.repo.path.tag_index if tags is None else tags
```
? Asking this as a question since on the one hand I like the `or` idiom but on the other I can see the API behaving the opposite of what would be expected if we happen to call this function with an empty list. Think for instance if the list is constructed during a script and ends up being empty.

tldahlgren replied:
🤔

alalazo replied:
The case I have in mind here is:
```python
# Assume this is the result of a computation and is empty
tags = []
result = get_tag_packages(tags, installed=False, skip_empty=True)
```
`result` will contain all the tags, while I assume the client expects it to be empty. It is also inconsistent with the documentation:
```
tags (list or None): list of tags of interest or None for all
```
because `None` is different from the empty list (but still false-ish)

tldahlgren replied:
You're absolutely right.  (I was side-tracked by something else so am now getting back to this.)

alalazo reviewed (changes requested):
This basically LGTM. I left a number of comments, mostly minor.

RikkiButler20 commented:
Could you update the documentation under Command Reference

tldahlgren committed:
Rebase cleanup; tags options changes; env-related refactor; and test cleanup

tldahlgren commented:
@alalazo I had problems rebasing that resulted in skipping a commit so I ended up trying to reproduce part of what was here.  Hopefully I've been successful at that *and* in addressing the concerns and suggestions you raised.

tldahlgren commented:
> Could you update the documentation under Command Reference

I expanded on the documentation that should get automatically picked up but I will check on this next chance I get.

tldahlgren committed:
Restore missing distinction when reporting installed packages

alalazo commented on lib/spack/spack/tag.py:
```diff
@@ -0,0 +1,141 @@
+def has_installed_packages():
+    """Returns True if there are installed packages, False otherwise."""
+    specs = spack.environment.get_installed_package_specs()
+    return len(specs) > 0
```
I think this is unnecessary. Aren't:
```python
if spack.environment.get_installed_package_specs():
    pass
```
and:
```python
if spack.tag.has_installed_packages():
    pass
```
equivalent? Can we remove this function?

tldahlgren replied:
> > Could you update the documentation under Command Reference
> 
> I expanded on the documentation that should get automatically picked up but I will check on this next chance I get.

@RikkiButler20 FYI. The Spack PR process already adds the new command to the table (see https://spack--26136.org.readthedocs.build/en/26136/command_index.html) and the documentation extracted from the comments (see https://spack--26136.org.readthedocs.build/en/26136/command_index.html#spack-tags).

Did you want more documentation than this?

alalazo commented on lib/spack/spack/tag.py:
```diff
@@ -0,0 +1,141 @@
+import spack.util.spack_json as sjson
+
+
+def get_installed_package_names():
```
Can we rename it to:
```suggestion
def _get_installed_package_names():
```
or inline it in the only place where it's used?

alalazo commented on lib/spack/spack/environment/environment.py:
```diff
@@ -102,6 +102,16 @@
 default_view_link = 'all'
 
 
+def get_installed_package_specs():
```
Nitpick. We can avoid `get_` and possibly `package` in the name:
```suggestion
def installed_specs():
```
When a client calls this function I think:
```python
spack.environment.installed_specs()
```
will still make sense. What do you think?

tldahlgren replied:
It does have a kind of symmetry with `uninstalled_specs`.  My concern was there is an `installed_specs` in `cmd/common/arguments.py` and it is used in a number of contexts as a variable but I will assume that won't cause problems.

alalazo reviewed (changes requested):
Thanks this looks great. I have just one comment on the output in https://github.com/spack/spack/pull/26136#discussion_r738961700 and a few very minor code comments.

tldahlgren committed:
Simplify interfaces per PR feedback

tldahlgren committed:
Resolve style errors
